src/components/CouponDetails/CouponBulkActions.jsx

```diff
@@ -1,56 +1,75 @@
-import React from 'react';
+import React, { useState, useMemo, useEffect } from 'react';
 import PropTypes from 'prop-types';
 
 import {
   Button, Form,
 } from '@edx/paragon';
-import { BULK_ACTION } from './constants';
+import { BULK_ACTION, COUPON_FILTERS, COUPON_FILTER_TYPES } from './constants';
+import { getBASelectOptions, getFirstNonDisabledOption } from './helpers';
 
 const CouponBulkActions = ({
-  value, options, disabled, onChange, handleBulkAction,
-}) => (
-  <>
-    <Form.Group className="flex-grow-1" controlId={BULK_ACTION.controlId} name={BULK_ACTION.name}>
-      <Form.Control
-        className="float-right"
-        floatingLabel={BULK_ACTION.label}
-        as="select"
-        value={value}
-        onChange={(e) => onChange(e.target.value)}
-        disabled={disabled}
+  handleBulkAction, selectedToggle, numSelectedCodes, numUnassignedCodes, couponAvailable, isTableLoading, hasTableData,
+}) => {
+  const isAssignView = selectedToggle === COUPON_FILTERS.unassigned.value;
+  const isRedeemedView = selectedToggle === COUPON_FILTERS.redeemed.value;
+
+  const options = useMemo(() => getBASelectOptions({
+    isAssignView,
+    isRedeemedView,
+    hasTableData,
+    couponAvailable,
+    numUnassignedCodes,
+    numSelectedCodes,
+  }), [numUnassignedCodes, numSelectedCodes, selectedToggle, hasTableData, couponAvailable]);
+
+  const [value, setValue] = useState(getFirstNonDisabledOption(options));
+  useEffect(() => {
+    setValue(getFirstNonDisabledOption(options));
```

Comment:
is there a potential that between the declaration of useState with the default value of `getFirstNonDisabledOption(options)` the value of that resolved method would change by this line? Ie why do we set the value to the same thing the we're defaulting to?

Reply:
This is actually the part that fixes the problem; it wasn't there in the parent component and would have been difficult to add for various reasons.

Basically this is saying, any time the `BulkActionSelectOptions` change, you should update the value here as well. Gonna make a small change to make that more obvious and keep them in sync.

One of the reasons for moving the code is that it was very hard to reason about in the parent component. It was split between three or four different methods across an 800+ line component. This is work I wanted to do in the refactor anyway but ran out of time for.

Reply:
And really, one of the main reasons is that it made it easier to test that my solution worked. I *could* test it solely via the parent component, but I have a lot more control over the inputs this way and I find it much easier to reason about.